google/cloud/aiplatform/data_source.py

```diff
@@ -0,0 +1,264 @@
+    To be used with the Dataset.create function to import the dataset at creation time.
+    """
+
+    def __init__(self, source_uri: str):
+        """
+        Args:
+            source_uris (Sequence[str]):
+                Required. Google Cloud Storage URI(-s) to the
+                input file(s). May contain wildcards. For more
+                information on wildcards, see
+                https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames.
+        """
+
+        # Perform validation of URI
+        if not (source_uri.startswith("bq://") or source_uri.startswith("bigquery://")):
+            raise ValueError(
+                "Expected source_uri to start with 'bq://' or 'bigquery://'"
+            )
+
+        self._source_uri = source_uri
+
+    @property
+    def metadata_schema_uri(self) -> str:
+        # Always return tabular as BQ only supports tabular
+        return schema.dataset.metadata.tabular
+
+    @property
+    def metadata(self) -> Dict:
+        return {"input_config": {"bigquery_source": {"uri": self._source_uri}}}
+
+
+# Do not instantiate this abstract class
+class GCSSourceValidating(DataSource):
+    """Abstract class that takes in GCS source_uri's and performs validation"""
+
+    def __init__(self, source_uris: [str]):
+        """
+        Args:
+            source_uris (Sequence[str]):
+                Required. Google Cloud Storage URI(-s) to the
+                input file(s). May contain wildcards. For more
+                information on wildcards, see
+                https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames.
+        """
+
+        # Perform validation of URI's
+        if not all([uri.startswith("gs://") for uri in source_uris]):
+            raise ValueError("Expected alls URI's in source_uris to start with 'gs://'")
+
+        self._source_uris = source_uris
+
+
+# Do not instantiate this abstract class
+class GCSDataSource(GCSSourceValidating, DataSource):
+    """Abstract class for GCS source_uri's that provides Dataset metadata"""
+
+    @property
+    def metadata(self) -> Dict:
+        return {"input_config": {"gcs_source": {"uri": self._source_uris}}}
+
+
+class GCSTabularDataSource(GCSDataSource):
+    """Class for a tabular dataset schema that provides Dataset metadata.
+
+    Used for CSV files on Google Cloud Storage.
+    To be used with the Dataset.create function to import the dataset at creation time.
+    """
+
+    @property
+    def metadata_schema_uri(self) -> str:
+        # Always return tabular as this config only supports tabular
+        return schema.dataset.metadata.tabular
+
+
+class DataImportable(abc.ABC):
+    """An abstract class that provides import_data_config for importing data to an existing database"""
+
+    @property
+    def import_data_config(self) -> ImportDataConfig:
+        raise NotImplementedError
+
+
+class GCSNonTabularImportDataSource(GCSSourceValidating, DataImportable):
```

Comment:
Importing a dataset after creation takes different parameters so ImportDataSources are separate classes that conform to `DataImportable`.

Reply:
This is a strange class because it seems to contain `metadata_schema_uri` and `metadata` as properties but they are not implemented.